mac/Keyman4MacIM/Keyman4MacIM/KMInputMethodEventHandler.m

```diff
@@ -42,22 +42,33 @@ - (instancetype)initWithLegacyMode:(BOOL)legacy clientSelectionCanChangeUnexpect
     return self;
 }
 
-// This is the public initializer.
-- (instancetype)initWithClient:(NSString *)clientAppId client:(id) sender {
-    self.senderForDeleteBack = sender;
-    // TODO: Pages and Keynote (and possibly lots of other undiscovered apps that are otherwise compliant
-    // with Apple's IM framework) have a problem in that if the user selects a different font (or other
-    // formatting) and then types a sequence that causes characters to be added to the document and then
-    // subsequently replaced, the replacement causes the formatting decision to be forgotten. This can be
-    // "fixed" by treating them as legacy apps, but it causes other problems.
-    BOOL legacy = ([clientAppId isEqual: @"com.github.atom"] ||
+/**
+ * Checks if the client app requires legacy input mode, first by checking the user defaults, if they exist,
+ * then, by our hard-coded list.
+ */
+- (BOOL)isClientAppLegacy:(NSString *)clientAppId {
+    NSArray *legacyAppsUserDefaults = [self.AppDelegate legacyAppsUserDefaults];
+
+    BOOL result = NO;
+
+    if(legacyAppsUserDefaults != nil) {
+        result = [self isClientAppLegacy:clientAppId fromArray:legacyAppsUserDefaults];
+    }
+
+    if(!result) {
```

Comment:
So, an app is marked as legacy if at least one of these two statements is true:

- The user's custom legacy listing says to treat it as legacy
- Our internal list says to treat it as legacy

----

So, if the user wants to remove an app from 'legacy' treatment, but the app is on the internal list, tough luck.  Aside from that, all other usage patterns are well-supported.

Is this accurate?

Reply:
Answer from team discussion:  yes, it is accurate, but we're fine moving forward with it this way; it's already an improvement.